File: kafka-0.7.0/kafka-0.7.0-core-java/src/main/java/kafka/utils/Pool.java
package kafka.utils;

import scala.Tuple2;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;

public class Pool<K, V> implements Iterable<Tuple2<K, V>> {
    
    private ConcurrentHashMap<K, V> pool = new ConcurrentHashMap<K, V>();
    
    public Pool() {
    }
    
    public Pool(ConcurrentHashMap<K, V> map) {
        for (K key : map.keySet()) {
            pool.put(key, map.get(key));
        }
    }
    
    public ConcurrentHashMap.KeySetView<K, V> keys() {
        return pool.keySet();
    }
    
    public Iterable<V> values() {
        return new ArrayList<V>(pool.values());
    }
    
    public int size() {
        return pool.size();
    }
    
    
    public void put(K k, V v) {
        pool.put(k, v);
    }
    
    
    public V putIfNotExists(K k, V v) {
        pool.putIfAbsent(k, v);
        return v;
    }
    
    public boolean contains(K id) {
        return pool.containsKey(id);
    }
    
    public V get(K key) {
        return pool.get(key);
    }
    
    public V remove(K key) {
        return (V) pool.remove(key);
    }
    
    
    public void clear() {
        pool.clear();
    }
    
    @Override
    public Iterator<Tuple2<K, V>> iterator() {
        return new Iterator<Tuple2<K, V>>() {
            private Iterator<Map.Entry<K, V>> iter = pool.entrySet().iterator();
            
            @Override
            public boolean hasNext() {
                return iter.hasNext();
            }
            
            @Override
            public Tuple2<K, V> next() {
                Map.Entry<K, V> n = iter.next();
                return new Tuple2<K, V>(n.getKey(), n.getValue());
            }
            
            @Override
            public void remove() {
            
            }
            
            @Override
            public void forEachRemaining(Consumer action) {
            
            }
        };
    }
}

File: kafka-0.7.0/kafka-0.7.0-core/src/main/scala/kafka/utils/Pool.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.utils

import java.util
import java.util.{ArrayList, Map}
import java.util.concurrent._

import collection.JavaConversions

class Pool[K,V] extends Iterable[(K, V)] {

  private val pool = new ConcurrentHashMap[K, V]
  
  def this(m: collection.Map[K, V]) {
    this()
    for((k,v) <- m)
      pool.put(k, v)
  }
  
  def put(k: K, v: V) = pool.put(k, v)
  
  def putIfNotExists(k: K, v: V) = pool.putIfAbsent(k, v)
  
  def contains(id: K) = pool.containsKey(id)
  
  def get(key: K): V = pool.get(key)
  
  def remove(key: K): V = pool.remove(key)
  
  def keys = JavaConversions.asScalaSet(pool.keySet())
  
  def values: Iterable[V] = 
//    JavaConversions.asIterable(new ArrayList[V](pool.values()))
    JavaConversions.asScalaBuffer(new ArrayList[V](pool.values()))
  
  def clear: Unit = pool.clear()
  
  override def size = pool.size
  
  override def iterator = new Iterator[(K,V)]() {
    
    private val iter: util.Iterator[Map.Entry[K, V]] = pool.entrySet.iterator
    
    def hasNext: Boolean = iter.hasNext
    
    def next: (K, V) = {
      val n = iter.next
      (n.getKey, n.getValue)
    }
    
  }
    
}
